### ISSUE TITLE
Avoid materialising unused Triton IF/LIF voltage sequences when store_v_seq=False

### ISSUE THREAD

**Issue by Magnushst** [fangwei123456/SpikingFlow#706]
## Environment

- SpikingJelly baseline: `2797c5575515b59d7f09a3d5b732d6d25e148d13` (`master`)
- OS: Windows 11 10.0.26200
- Python: 3.11.15
- PyTorch: 2.7.1+cu128
- Triton: 3.3.1
- PyTorch CUDA runtime: 12.8
- Installed CUDA toolkit: 12.1 (`CUDA_PATH`)
- NVIDIA driver: 596.36
- GPU: NVIDIA GeForce RTX 4070 GPU, 8 GB, compute capability 8.9

## Problem

`IFNode` and `LIFNode` default to `store_v_seq=False`, which means callers only retain the final membrane potential. The standard Triton multi-step path nevertheless allocates a full `v_seq` tensor with the same `[T, N, ...]` shape as the input, writes every timestep into it, and then uses its final row to update the node state. Training also presents a full voltage-gradient sequence to the backward kernel even though only the final state is externally visible in this mode. This makes the default behaviour inconsistent with the intent of `store_v_seq=False` and adds allocation and memory traffic proportional to the full sequence length.

## Reproduction

The benchmark script is available on the proposed
[`perf/triton-last-state`](https://github.com/Magnushst/spikingjelly/tree/perf/triton-last-state)
branch. It uses a fixed seed of `20260717`, CUDA events, 5 warm-ups, 25 repetitions,
10 iterations per repetition, and `torch.cuda.max_memory_allocated()`:

```bash
python benchmark/benchmark_triton_last_state.py --output optimised.json
```

For a clean baseline/working-tree comparison:

```bash
git worktree add ../spikingjelly-baseline 2797c5575515b59d7f09a3d5b732d6d25e148d13
SJ_BENCH_SOURCE_ROOT=../spikingjelly-baseline \
  python benchmark/benchmark_triton_last_state.py --output baseline.json
python benchmark/benchmark_triton_last_state.py --output optimised.json
```

## Baseline evidence

Peak allocated memory for the default `store_v_seq=False` path:

| Mode | Shape / dtype | Baseline | Proposed | Reduction |
|---|---:|---:|---:|---:|
| Inference | `T=16, N=4096`, FP32 | 0.531 MiB | 0.281 MiB | 47.1% |
| Inference | `T=32, N=16381`, FP16 | 2.062 MiB | 1.062 MiB | 48.5% |
| Inference | `T=65, N=16381`, FP32 | 8.249 MiB | 4.187 MiB | 49.2% |
| Training | `T=16, N=4096`, FP32 | 1.532 MiB | 1.298 MiB | 15.3% |
| Training | `T=32, N=16381`, FP16 | 6.063 MiB | 5.095 MiB | 16.0% |
| Training | `T=65, N=16381`, FP32 | 24.498 MiB | 20.499 MiB | 16.3% |

IF and LIF produced the same peak-allocation figures. Four counter-ordered latency trials were noisy and mixed on this laptop, so this should be presented as a memory optimisation rather than as an end-to-end latency improvement.

For `T=65, N=16381`, FP32 IF training, a 10-iteration PyTorch Profiler capture showed:

- Forward custom-op device allocation: 127,779,840 to 85,841,920 bytes across 10 iterations.
- Forward kernel device time: 219.452 to 199.262 microseconds across 10 calls.
- Backward kernel device time: 219.868 to 132.890 microseconds across 10 calls.
- `aten::copy_` count: 20 to 10; the 10 device-to-device copies associated with retaining the final row disappeared.

## Proposed change

- Thread the existing node-level `store_v_seq` flag into standard IF/LIF Triton operators.
- Allocate a final-state-sized voltage output when the flag is false.
- Skip per-timestep voltage stores and write only the final state once.
- Seed the backward recurrence from the final-state gradient and skip per-timestep voltage-gradient loads in this mode.
- Preserve the existing full-sequence path when `store_v_seq=True`.
- Leave PLIF and mixed-precision neuron paths unchanged because their current contracts and intermediate requirements differ.

## Compatibility

- `IFNode`/`LIFNode` public behaviour and defaults are unchanged.
- The low-level `multistep_if`/`multistep_lif` helpers gain an optional `store_v_seq=True` argument, preserving existing callers and return behaviour by default.
- Torch, CuPy, PLIF, CPU, and mixed-precision paths are unchanged.
- No dependency or minimum-version change is required.
- The boolean is a Triton compile-time specialisation and an autotune-key component, so standard IF/LIF kernels can have separate full-sequence and final-state variants.

## Validation

- Compared full-sequence and final-state paths for IF and LIF.
- Covered inference and training, the default static-loop path (`T=1` and `T=7`), and the dynamic-loop path (`T=65`).
- Covered FP32, FP16, and BF16, hard and soft reset, and detach-reset.
- Compared spikes, final membrane state, and input gradients with dtype-appropriate tolerances.
- Exercised `torch.compile` with Inductor and checked fake-tensor output shapes.
- Benchmarked power-of-two and non-power-of-two feature counts with raw samples and peak allocation.

The complete relevant compile-compatibility module finished with 78 tests passed and 3 expected skips requiring a second CUDA device.

## Scope question

Would maintainers accept a focused PR that makes the standard Triton IF/LIF path honour `store_v_seq=False` at allocation and kernel level, while keeping full-sequence, mixed-precision, PLIF, and non-Triton behaviour unchanged?

**Comment by AllenYolk** [comment]
Thank you for the detailed investigation, benchmarks, and validation. The proposed optimization makes sense: when `store_v_seq=False`, avoiding the full `v_seq` allocation and per-timestep writes better matches the intended semantics and reduces unnecessary memory usage and traffic.

We noticed that you have already implemented the proposed changes in your fork on the [`perf/triton-last-state`](https://github.com/Magnushst/spikingjelly/tree/perf/triton-last-state) branch. Could you please submit a pull request to the main repository? We would be happy to review the implementation and tests there.

**Comment by Magnushst** [comment]
> Thank you for the detailed investigation, benchmarks, and validation. The proposed optimization makes sense: when `store_v_seq=False`, avoiding the full `v_seq` allocation and per-timestep writes better matches the intended semantics and reduces unnecessary memory usage and traffic.
> 
> We noticed that you have already implemented the proposed changes in your fork on the [`perf/triton-last-state`](https://github.com/Magnushst/spikingjelly/tree/perf/triton-last-state) branch. Could you please submit a pull request to the main repository? We would be happy to review the implementation and tests there.

Thank you for confirming the scope. PR #709 is now open for review: https://github.com/fangwei123456/spikingjelly/pull/709
